File: runtime/erlang/src/runtime_http.erl
%% Shared smithy-beam Erlang HTTP request/response helper module.
-module(runtime_http).
-include("runtime_types.hrl").
-export([
    dispatch/2,
    dispatch/3,
    with_retry/2
]).

%% @doc Dispatch an HTTP request using the client config default HTTP adapter.
-spec dispatch(#{binary() => term()}, http_request()) ->
    {ok, http_response()} | {error, term()}.
dispatch(Config, Request) ->
    HttpClient = maps:get(http_client, Config, runtime_http_client_httpc),
    dispatch(HttpClient, Config, Request).

%% @doc Dispatch an HTTP request through a specific HTTP client module.
-spec dispatch(module(), #{binary() => term()}, http_request()) ->
    {ok, http_response()} | {error, term()}.
dispatch(HttpClient, Config, Request) ->
    ClientReq = runtime_http_client:build_request(Config, Request),
    HttpClient:request(ClientReq).

%% @doc Invoke Fun with exponential backoff when a retryable error is returned.
-spec with_retry(fun(() -> term()), map()) -> term().
with_retry(Fun, Opts) ->
    Max = maps:get(max_attempts, Opts, 3),
    Base = maps:get(base_delay_ms, Opts, 100),
    ShouldRetry = maps:get(should_retry, Opts, fun(_) -> false end),
    with_retry(Fun, Max, Base, 1, ShouldRetry).

-spec with_retry(fun(() -> term()), non_neg_integer(), non_neg_integer(), pos_integer(), fun(
    (term()) -> boolean()
)) ->
    term().
with_retry(Fun, 0, _, _, _) ->
    Fun();
with_retry(Fun, Attempts, Base, N, ShouldRetry) ->
    case Fun() of
        {ok, _} = Ok ->
            Ok;
        {error, _} = Err ->
            case ShouldRetry(Err) of
                true when Attempts > 1 ->
                    timer:sleep(trunc(Base * math:pow(2, N - 1))),
                    with_retry(Fun, Attempts - 1, Base, N + 1, ShouldRetry);
                _ ->
                    Err
            end
    end.

File: runtime/elixir/lib/runtime_http.ex
defmodule RuntimeHttp do
  @moduledoc "Generated HTTP dispatcher for Smithy service clients. Uses Req."

  alias RuntimeHttpClient
  alias RuntimeTypes.HttpRequest
  alias RuntimeTypes.HttpResponse

  @spec dispatch(map(), HttpRequest.t()) :: {:ok, HttpResponse.t()} | {:error, term()}
  def dispatch(config, req) do
    http_client = Map.get(config, :http_client, RuntimeHttpClient.Req)
    dispatch(http_client, config, req)
  end

  @spec dispatch(module(), map(), HttpRequest.t()) ::
          {:ok, HttpResponse.t()} | {:error, term()}
  def dispatch(http_client, config, req = %HttpRequest{}) do
    client_req = RuntimeHttpClient.build_request(config, req)
    http_client.request(client_req)
  end

  @doc "Invokes fun with exponential backoff when a retryable error is returned."
  @spec with_retry((-> term()), keyword()) :: term()
  def with_retry(fun, opts) do
    max_attempts = Keyword.get(opts, :max_attempts, 3)
    base_delay_ms = Keyword.get(opts, :base_delay_ms, 100)
    should_retry = Keyword.get(opts, :should_retry, fn _ -> false end)
    with_retry(fun, max_attempts, base_delay_ms, 1, should_retry)
  end

  defp with_retry(fun, 0, _base, _n, _should_retry), do: fun.()

  defp with_retry(fun, attempts, base, n, should_retry) do
    case fun.() do
      {:ok, _} = ok ->
        ok

      {:error, _} = err ->
        if should_retry.(err) and attempts > 1 do
          Process.sleep(trunc(base * :math.pow(2, n - 1)))
          with_retry(fun, attempts - 1, base, n + 1, should_retry)
        else
          err
        end
    end
  end
end
